Support getChildContext() in shallow render

minznerjosh opened the pull request:
Fixes #664

This PR adds support for `getChildContext()` when shallow-rendering. I've decided to only implement parent-based context, not owner-based context (as I'm not exactly sure what that behavior is, tbh) which means I've excluded `react@0.13` from this implementation. (Hope that's okay!)

In `react@<=15`, the react shallow renderer actually calls `getChildContext()` for us! This means, as is the case with `shouldComponentUpdate()`, we just spy on that method's return value.

In `react@>=16`, `getChildContext()` is _not_ called (understandable—it's a deprecated API), so we have to call it ourselves.

P.S. Sorry the diff is a little ugly. I ended up messing with the nesting of some conditionals to avoid duplicated checks.

minznerjosh commented:
P.S. I'm pretty sure the build will succeed once is #1968 is in.

ljharb commented:
@minznerjosh I've rebased this and split up the commits a bit more atomically; it'd be great if you could add some tests to each commit/change :-)

minznerjosh commented:
@ljharb you’ve got it! I’ll take care of it when I get back from vacation on Wednesday. Thanks for taking a look at this!

minznerjosh commented:
@ljharb I added tests to every commit except ```[enzyme-adapter-react-*] [new] add `legacyContextMode`, and `getChildContext` lifecycle method``` (which is just a configuration change.) This is ready for another round of review!

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -180,6 +185,55 @@ function isPureComponent(instance) {
+    }
+  });
+  if (typeof renderer.checkPropTypes === 'function') {
+    renderer.checkPropTypes(Component.childContextTypes, childContext, 'child context', hierarchy);
```
are childContextTypes new in React 16?

minznerjosh replied:
Not at all! I did a bit of hunting and the [first instance of `childContextTypes` I can find in the source](https://github.com/facebook/react/blob/v0.9.0/src/core/ReactCompositeComponent.js#L968-L996) is in `v0.9.0`. I believe this means it's been part of the context API since the beginning.

ljharb replied:
Then why wouldn't every adapter, not just the 16+ ones, implement `checkPropTypes`? or, why wouldn't we implement it in enzyme-adapter-utils instead of in each adapter?

minznerjosh replied:
The <16 shallow renderers call `getChildContext()` for us and implement this logic themselves. So there's no need for the <16 renderers to implement `checkPropTypes` because nothing uses them.

However, if you think the method should be there just for the sake of consistency, I can add it (and refactor it into enzyme-adapter-utils.)

ljharb replied:
I'm confused; this problem (enzyme doesn't provide proper child context) definitely predates react 16 - since the shallow renderer only renders one level, i'm not sure how it would take child context into account?

ljharb replied:
Gotcha. So we still have to do work for every react version, only *more* work in 16+.

minznerjosh replied:
Exactly!

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -180,6 +185,55 @@ function isPureComponent(instance) {
+  }
+  const childContext = instance.getChildContext();
+  Object.keys(childContext).forEach((key) => {
+    if (!(key in Component.childContextTypes)) {
```
are we using `in` here because React does? it seems like this should be an own property check.

minznerjosh replied:
Yeah I tried to match react as much as possible [and they are using `in`](https://github.com/facebook/react/blob/1454a8be03794f5e0b23a7e7696cbbbdcf8b0f5d/packages/react-dom/src/server/ReactPartialRenderer.js#L633).

ljharb replied:
Let's add a comment then, above line 193, linking to https://github.com/facebook/react/blob/1454a8be03794f5e0b23a7e7696cbbbdcf8b0f5d/packages/react-dom/src/server/ReactPartialRenderer.js#L630-L637

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -180,6 +185,55 @@ function isPureComponent(instance) {
+
+function privateSetChildContext(adapter, wrapper, instance, renderedNode, getChildContextSpy) {
+  const renderer = wrapper[RENDERER];
+  // We only support parent-based context.
```
what does this mean for react 0.13? is everything the same as before the PR?

minznerjosh replied:
Correct!

ljharb replied:
Let's file a new issue after this is merged, then, to track adding support for that.

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -180,6 +185,55 @@ function isPureComponent(instance) {
+function privateSetChildContext(adapter, wrapper, instance, renderedNode, getChildContextSpy) {
+  const renderer = wrapper[RENDERER];
+  // We only support parent-based context.
+  if (adapter.options.legacyContextMode !== 'parent') { return; }
```
what about using older versions of the react 0.13 adapter, that won't have this property?

minznerjosh replied:
It should behave exactly as before `undefined !== 'parent'` is `true` so this function will just early-return. But I will add some tests to increase our confidence.

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -197,6 +251,8 @@ class ShallowWrapper {
       }
 
+      const getChildContextSpy = lifecycles.getChildContext.calledByRenderer
+        ? spyOnGetChildContextInitialRender(nodes, adapter) : null;
```
if the entire ternary can't fit on one line, then the `?` and `:` need to both be on their own lines

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -509,17 +577,28 @@ class ShallowWrapper {
+            getChildContextSpy = spyMethod(instance, 'getChildContext');
+          }
+        }
+        if (!shouldComponentUpdateSpy && isPureComponent(instance)) {
```
I believe PureComponents completely ignore any shouldComponent methods, so if it's a PureComponent, it should always fall into this branch

minznerjosh replied:
Interesting! Based on my reading of the [current implementation in master](https://github.com/airbnb/enzyme/blob/c7282ebc52aafd3442f7b52dde75354bae973674/packages/enzyme/src/ShallowWrapper.js#L379-L392), we don't do the `isPureComponent` check if `instance.shouldComponentUpdate` is a function (as per the first `if ()` condition.

So this implementation more closely matches the implementation in master. However, perhaps what's in master is wrong. In that case, would you like me to fix it here, or should that fix happen separately?

ljharb replied:
Well, first, let's double check me by confirming what React does. If React doesn't match master, then we probably should get in a PR to fix that prior to this one.

minznerjosh replied:
It looks like master (and this PR) has the correct behavior. [React first checks if `shouldComponentUpdate` is a function, and _then_ checks if the component is a `PureComponent`](https://github.com/facebook/react/blob/4846809370737f13c0465ac7b83764b408c6e732/packages/react-reconciler/src/ReactFiberClassComponent.js#L259-L284).

ljharb commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -1411,7 +1493,15 @@ class ShallowWrapper {
+        ...options,
+        context: {
+          ...this[OPTIONS].context,
+          ...options.context,
```
i wonder if we should instead consider the `context` option as *overriding* the proper child context - ie, only do the correct thing when a context option isn't supplied? (or alternatively, not pass forward the context option?

minznerjosh replied:
I like the idea of the `context` option being an override!

minznerjosh commented:
@ljharb 

* Fixed the formatting of that ternary
* Added tests that simulate an old adapter
* Made it so that the `context` option overrides a parent's context (and added a test)

This is ready for another round of review!

minznerjosh commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -180,6 +185,63 @@ function isPureComponent(instance) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `${componentName}.getChildContext(): childContextTypes `
+      + 'must be defined in order to use getChildContext().',
```
[React does warn](https://github.com/facebook/react/blob/1454a8be03794f5e0b23a7e7696cbbbdcf8b0f5d/packages/react-dom/src/server/ReactPartialRenderer.js#L639-L646)! I added a comment with a link to the react impl.

minznerjosh commented on packages/enzyme/src/ShallowWrapper.js:
```diff
@@ -1411,7 +1502,15 @@ class ShallowWrapper {
+        context: {
+          ...this[OPTIONS].context,
+          ...this[ROOT][CHILD_CONTEXT],
+          ...options.context,
```
That wouldn't quite work, since we're relying on passing `options.context` in order to combine contexts as we `dive()` through multiple levels of providers.

I've updated the code to this:
```javascript
context: options.context || {
  ...this[OPTIONS].context,
  ...this[ROOT][CHILD_CONTEXT],
}
```

So now `options.context` (passed to `dive()`) always takes precedence, but context is still combined when `dive()`ing multiple providers.

ljharb replied:
hmm. I'm not sure that's what we want. if it's not in the child context, it shouldn't be passed to the new child.

ljharb replied:
ahhhh ok - then yes, the larger object should indeed be passed down, as long as the component only sees the filtered context. Thanks for explaining.

minznerjosh replied:
My pleasure!

minznerjosh commented:
@ljharb 

* Should we implement `checkPropTypes` in all the adapters, even though it's only used in the >=16 adapters?
* Removed multi-line strings
* Added comment to explain warning if `childContextTypes` is not defined, with link to react src
* Added comment to explain checking of propTypes implementation, with link to react src
* Adjusted implementation of `dive()` so that `options.context` (passed to `dive()`) overrides anything else.

This is ready for another round!

ljharb commented:
@minznerjosh is there any chance the functionality is actually accessible inside one of the react packages, without copy/paste?

minznerjosh commented:
@ljharb Unfortunately, I don't think so. 🙁Both pieces of logic live in the `react-reconciler` package [here](https://github.com/facebook/react/blob/10a7a5b5ced683125d68584a78084faac3197846/packages/react-reconciler/src/ReactFiberClassComponent.js#L492-L499) and [here](https://github.com/facebook/react/blob/10a7a5b5ced683125d68584a78084faac3197846/packages/react-reconciler/src/ReactFiberContext.js#L149-L213), but neither are public exports, and the public exports that use those functions (like [`updateContainer`](https://github.com/facebook/react/blob/10a7a5b5ced683125d68584a78084faac3197846/packages/react-reconciler/src/ReactFiberReconciler.js#L283-L299)) do too much to be useful for us.

minznerjosh commented:
P.S. Just wanna ping [the conversation happening over here](https://github.com/airbnb/enzyme/pull/1971#discussion_r249040520) as that thread still needs to be closed. 😊

ljharb commented:
@minznerjosh what about https://unpkg.com/prop-types@15.6.2/checkPropTypes.js ?

minznerjosh commented:
@ljharb Can we live with the duplication of react logic here? If so, anything else blocking?

ljharb commented:
It's not ideal, but we can live with it :-) I'm reviewing and making some tweaks locally now.

ljharb commented:
Thanks for fixing this long-standing bug!
